### ISSUE TITLE
don't call attach callback in lm_ctx_create() for LM_TRANS_SOCK

### ISSUE THREAD

**Issue by changpe1** [nutanix/muser#61]
```
      // Attach to the muser control device.
-    lm_ctx->conn_fd = transports_ops[dev_info->trans].attach(lm_ctx);
     if ((dev_info->flags & LM_FLAG_ATTACH_NB) == 0) {
+        lm_ctx->conn_fd = transports_ops[dev_info->trans].attach(lm_ctx);
```

The socket conn_fd can only be valid in  `lm_ctx_try_attach()` function
```
@@ -2376,12 +2376,19 @@ int
 lm_ctx_try_attach(lm_ctx_t *lm_ctx)
 {
     assert(lm_ctx != NULL);
+    int ret;

     if ((lm_ctx->flags & LM_FLAG_ATTACH_NB) == 0) {
         errno = EINVAL;
         return -1;
     }
-    return transports_ops[lm_ctx->trans].attach(lm_ctx);
+
+    ret = transports_ops[lm_ctx->trans].attach(lm_ctx);
+    if (ret == -1) {
+        return -1;
+    }
+    lm_ctx->conn_fd = ret;
+    return 0;
 }
```

**Comment by tmakatos** [comment]
```
      // Attach to the muser control device.
     lm_ctx->conn_fd = transports_ops[dev_info->trans].attach(lm_ctx);
```
This calls `open_sock()` and if the client is not ready then `accept()` fails and returns `-1`, which is then assigned to `lm_ctx->conn_fd`. `lm_ctx->conn_fd` is already initailized to `-1`. I don't see where the problem is with the existing implementation.

**Comment by changpe1** [comment]
> ```
>       // Attach to the muser control device.
>      lm_ctx->conn_fd = transports_ops[dev_info->trans].attach(lm_ctx);
> ```
> 
> This calls `open_sock()` and if the client is not ready then `accept()` fails and returns `-1`, which is then assigned to `lm_ctx->conn_fd`. `lm_ctx->conn_fd` is already initailized to `-1`. I don't see where the problem is with the existing implementation.

but the conn_fd is always -1 then, because lm_ctx_try_attach doesn't pass the correct file descriptor to it.

**Comment by tmakatos** [comment]
> but the conn_fd is always -1 then, because lm_ctx_try_attach doesn't pass the correct file descriptor to it.
`lm_ctx_try_attach()` passes `lm_ctx->fd` to `accept()`, which is initialized in `lm_ctx_ctreate()`:
```
    if (transports_ops[dev_info->trans].init != NULL) {
        err = transports_ops[dev_info->trans].init(lm_ctx);
        if (err < 0) {
            goto out;
        }
        lm_ctx->fd = err;
    }
```

**Comment by changpe1** [comment]
> > but the conn_fd is always -1 then, because lm_ctx_try_attach doesn't pass the correct file descriptor to it.
> 
> `lm_ctx_try_attach()` passes `lm_ctx->fd` to `accept()`, which is initialized in `lm_ctx_ctreate()`:
> 
> ```
>     if (transports_ops[dev_info->trans].init != NULL) {
>         err = transports_ops[dev_info->trans].init(lm_ctx);
>         if (err < 0) {
>             goto out;
>         }
>         lm_ctx->fd = err;
>     }
> ```

maybe lm_ctx->conn_fd = conn_fd is required in open_sock() function.

**Comment by tmakatos** [comment]
> maybe lm_ctx->conn_fd = conn_fd is required in open_sock() function.

Why? `lm_ctx->fd` is the file descriptor of the socket and `lm_ctx->conn_fd` is the file descriptor of the connection with the client. What is the problem with the current code?

**Comment by changpe1** [comment]
> > maybe lm_ctx->conn_fd = conn_fd is required in open_sock() function.
> 
> Why? `lm_ctx->fd` is the file descriptor of the socket and `lm_ctx->conn_fd` is the file descriptor of the connection with the client. What is the problem with the current code?

Polling lm_ctx->conn_fd will return errno with Bad file descriptor, because lm_ctx->conn_fd is -1, following code can fix the issue

@@ -2376,12 +2376,19 @@ int
 lm_ctx_try_attach(lm_ctx_t *lm_ctx)
 {
     assert(lm_ctx != NULL);
+    int ret;

     if ((lm_ctx->flags & LM_FLAG_ATTACH_NB) == 0) {
         errno = EINVAL;
         return -1;
     }
-    return transports_ops[lm_ctx->trans].attach(lm_ctx);
+
+    ret = transports_ops[lm_ctx->trans].attach(lm_ctx);
+    if (ret == -1) {
+        return -1;
+    }
+    lm_ctx->conn_fd = ret;
+    return 0;
 }

**Comment by tmakatos** [comment]
Ah, I just realized what the problem is, this commit changed it :
```
commit 435acf689f4ec5c199fff9c3fe4057de8bba02ce (HEAD -> vfio-user, upstream/vfio-user)
Author: Swapnil Ingle <swapnil.ingle@nutanix.com>
Date:   Fri Oct 16 04:53:14 2020 -0400

    libmuser: Try to attach device for non-block ctx

    Quick try to attach device while lm_create_ctx with LM_FLAG_ATTACH_NB

    Signed-off-by: Swapnil Ingle <swapnil.ingle@nutanix.com>

diff --git a/lib/libmuser.c b/lib/libmuser.c
index 24238b8..14d8b9e 100644
--- a/lib/libmuser.c
+++ b/lib/libmuser.c
@@ -2479,15 +2479,14 @@ lm_ctx_create(const lm_dev_info_t *dev_info)
     err = 0;

     // Attach to the muser control device.
+    lm_ctx->conn_fd = transports_ops[dev_info->trans].attach(lm_ctx);
     if ((dev_info->flags & LM_FLAG_ATTACH_NB) == 0) {
-        lm_ctx->conn_fd = transports_ops[dev_info->trans].attach(lm_ctx);
```
We shouldn't try to attach when `LM_FLAG_ATTACH_NB` is set, this solves the problem. Also, attempting to attach during `lm_ctx_create` might fail but we swallow the error, so the user of `lm_ctx` create doesn't know whether or not the device has been attached. @swapnili is there a reason why this behaviour changed?

**Comment by changpe1** [comment]
Yeah, but this isn't enough, we still need to set lm_ctx->conn_fd to the correct value in somewhere, maybe set it in open_sock() is better

**Comment by swapnili** [comment]
@tmakatos as we discussed previously you said there was no valid reason to not to try attempt attach in lm_ctx_create().
Also in the code following comment says the same.

 /*
     * Attaching to the transport is non-blocking. The library will attempt to
     * attach during context creation time and if attaching fails because the
     * transport is not ready then the error will be ignored. The caller must
     * then manually call lm_try_ctx_attach, which is non-blocking, as many
     * times as necessary.
     */
#define LM_FLAG_ATTACH_NB  (1 << 0) 

If we decide to change this behaviour we must also fix this comment, its misleading.

**Comment by tmakatos** [comment]
> @tmakatos as we discussed previously you said there was no valid reason to not to try attempt attach in lm_ctx_create().

@swapnili I completely forgot about this discussion. Apparently there is a problem with that so we should revert the patch and fix the comment.

@changpe1 if the `libmuser: Try to attach device for non-block ctx` patch is reverted, there shouldn't be a problem leaving `lm_ctx->conn_fd` uninitialized since after `lm_ctx_create()` the user _must_ repeatedly call `lm_ctx_try_attach()` until it's succesful; you mustn't call `lm_ctx_poll()` before that. I suppose we could initialize `lm_ctx->conn_fd` to -1 in `lm_ctx_create()` for clarity but stricly speaking it shouldn't be required. Right?

**Comment by changpe1** [comment]
Thank you, lm_ctx_try_attach() will return the correct connection fd, but the caller(SPDK) doesn't need to save this value, so libmuser should save it to lm_ctx->conn_fd, right?

**Comment by swapnili** [comment]
If attach fails in lm_ctx_create() then lm_ctx->conn_fd = -1. Why not caller can call lm_ctx_try_attach() only if lm_ctx->conn_fd == -1 and once conn_fd is correctly initialised then call  lm_ctx_poll() on it.

**Comment by swapnili** [comment]
Ok lm_ctx is opaque. So caller do not know if conn_fd is initialised or not. Then I will revert that patch and also update the comment.

**Comment by tmakatos** [comment]
>  but the caller(SPDK) doesn't need to save this value, so libmuser should save it to lm_ctx->conn_fd, right?

Correct, libmuser saves it in `lm_ctx->conn_fd`, the caller doesn't need to store it. The caller only needs to check that `lm_ctx_try_attach()` succeeded, which means it returned >= 0.

> Ok lm_ctx is opaque. So caller do not know if conn_fd is initialised or not. Then I will revert that patch and also update the comment.

Exactly.

**Comment by tmakatos** [comment]
Fixed by b9ae89c944770e75e74da946e6bd07932e77af23
